### ISSUE TITLE
docker "/course: Permission denied" when trying to mount my course 

### ISSUE THREAD

**Issue by cbm755** [PrairieLearn/PrairieLearn#4167]
I'm having trouble getting docker/podman to access my local course.  I'm sure this is fixable with `--userns` or something like that.  Will report back once I figure it out!

1.  Successful base case: Ubuntu 18.04 using a bare-metal file system:
```
# docker run -it --rm  -p 3000:3000 -v /honest_actual_disc/tmp:/course prairielearn/prairielearn touch /course/hello.txt
<success>
```

2.  Non-root user running Podman (Fedora drop-in Docker without root daemon stuff)
```
# whoami
cbm
# podman run -it --rm  -p 3000:3000 -v /home/cbm/tmp2:/course prairielearn/prairielearn ls /course
ls: cannot open directory /course: Permission denied
```

3.  Ubuntu 18.04 Docker but with a *shudder* NFS filesystem:
```
$ docker run -it --rm  -p 3000:3000 -v /zfs/users/cbm/tmp2:/course prairielearn/prairielearn ls /course
<success>
$ docker run -it --rm  -p 3000:3000 -v /zfs/users/cbm/tmp2:/course prairielearn/prairielearn touch /course/hello.txt
touch: cannot touch ‘/course/hello.txt’: Permission denied
```
(this is because root on a local machine is not root-on-nfs)

**Comment by echuber2** [comment]
I don't have podman installed to try it out, but you might be able to use `--uidmap` and `--gidmap` to remap 0 to your current local user uid and gid (e.g., 1000 on Ubuntu typically).

If that doesn't work for some reason, you might try rebuilding the image for local testing purposes. If I rebuild the Docker with this, I can get user 1000 to touch `/course/hello.txt`, but it still will break just about everything else:

``` dockerfile
FROM prairielearn/prairielearn
USER 0
RUN chmod -R a+w /course
```
Instead, you could rebuild the image recursively remapping all 0:0 stuff to 1000:1000 or whatever you need. For example, see the top of the example here: https://github.com/PrairieLearn/PrairieLearn/issues/3170 But, `uidmap` and `gidmap` sound like they can do this much more easily.

**Comment by cbm755** [comment]
TL;DR: if cannot read mount point, try adding `:z`.  Rest of what I'm thinking about is a dupe of #3170, closing.

- - - -

### Brain dump

I remembered/relearned that sometimes I need a `:z` on the end of my mounts: `-v /zfs/users/cbm/tmp2:/course:z` (mumble mumble selinux something something).

I gave up on writing to mounts on rootless podman: I agree it should work with the `--uidmap` but its a bit mixed up (at least for me) with `/etc/subuid`: I could not get the mount to be owned by anything except uid 65534 :(

Podman-as-root: this works fine.  Writes files as root which is annoying.  Can pass `--user 1000` but seems likely the image is not functional (not really tested).

I need to sort out write permissions stuff for a different project so if I can help more later in #3170 I will.

- - - - 

BUT I realized that for what I want (local preview of Questions before pushing MR/PR to GitLab/Github) that readonly access is good enough (and likely even preferable!).  For this `:z` seems to work (and for some reason, now it also works without `:z`, although as per my original comment I don't think it did.
